language/src/main.rs

```diff
@@ -65,93 +68,231 @@ impl HacspecErrorEmitter for Session {
+                    })
+                    .collect(),
+                span,
+            }
+        }
+    };
+
+    /////////////////////////////////
+    // Start of actual translation //
+    /////////////////////////////////
+
+    let external_data = |imported_crates: &Vec<rustspec::Spanned<String>>| {
+        queries.global_ctxt().unwrap().peek_mut().enter(|tcx| {
+            hir_to_rustspec::retrieve_external_data(&compiler.session(), &tcx, imported_crates)
+        })
+    };
+
+    let krate = match ast_to_rustspec::translate(&compiler.session(), &krate, &external_data) {
+        Ok(krate) => krate,
+        Err(_) => {
+            compiler
+                .session()
+                .err("unable to translate to Hacspec due to out-of-language errors");
+            return Compilation::Stop;
+        }
+    };
+
+    let krate = match name_resolution::resolve_crate(
+        &compiler.session(),
+        krate,
+        &external_data,
+        new_top_ctx,
+    ) {
+        Ok(krate) => krate,
+        Err(_) => {
+            compiler
+                .session()
+                .err("found some Hacspec name resolution errors");
+            return Compilation::Stop;
+        }
+    };
+
+    let krate = match typechecker::typecheck_program(&compiler.session(), &krate, new_top_ctx) {
+        Ok(krate) => krate,
+        Err(_) => {
+            compiler
+                .session()
+                .err("found some Hacspec typechecking errors");
+            return Compilation::Stop;
+        }
+    };
+    let imported_crates = name_resolution::get_imported_crates(&krate);
+    let imported_crates = imported_crates
+        .into_iter()
+        .filter(|(x, _)| x != "hacspec_lib")
+        .map(|(x, _)| x)
+        .collect::<Vec<_>>();
+    println!(
+        " > Successfully typechecked{}",
+        if imported_crates.len() == 0 {
+            ".".to_string()
+        } else {
+            format!(
+                ", assuming that the code in crates {} has also been Hacspec-typechecked",
+                imported_crates.iter().format(", ")
+            )
+        }
+    );
+
+    match &callback.output_directory {
+        None => return Compilation::Stop,
+        Some(file_str) => {
+            let original_file = Path::new(file_str);
+            let extension = &callback.output_type;
+
+            let file = match extension.clone().as_str() {
+                "fst" | "ec" | "json" => {
+                    // Compute file name as output directory with crate local path (krate_path)
+                    (original_file)
+                        .join(Path::new(
+                            (krate_path
+                                .clone()
+                                .to_title_case()
+                                .replace(" ", ".")
+                                + "." + extension)
+                                .as_str(),
+                        ))
+                }
+                "v" => {
```

Comment:
I guess you renamed the examples to get nice names here? 🤔
Can't you just prefix "Hacspec_" here for coq as you did before?

Reply:
Seems like a strange requirement that we remove hacspec from the file name, because of the way the _Coqproject files is set up. I think it is better to have something more general, and then change the _Coqproject file instead.

Reply:
Yep, see my other comment. There should be no assumption on file or crate names in here. So whatever style you want to have for Coq you should craft here. (Or don't do anything and change file names on the outside.)

Reply:
Yes, that was why I opted away from having anything case on "Hacspec_", and just have the file names in a format that coq accepts, eg. "_" instead of ".".